src/System.Buffers/src/System/Buffers/ArrayPool.cs

```diff
@@ -0,0 +1,97 @@
+    /// </summary>
+    public abstract class ArrayPool<T>
+    {
+        /// <summary>The default number of arrays that are available for rent.</summary>
+        public const int DefaultNumberOfArraysPerBucket = 50;
+        /// <summary>The default length of each Rent'able array; equal to 1MB in bytes</summary>
+        public const int DefaultArrayLength = 1024000;
+
+        private static ArrayPool<T> _sharedInstance = null;
+
+        /// <summary>
+        /// Retrieves a shared instance pool that is thread-safe and can be used by 
+        /// multiple components of the same system.
+        /// </summary>
+        public static ArrayPool<T> Shared
+        {
+            get
+            {
+                if (Volatile.Read(ref _sharedInstance) == null)
+                {
+                    Interlocked.CompareExchange(ref _sharedInstance, new DefaultArrayPool<T>(DefaultArrayLength, DefaultNumberOfArraysPerBucket), null);
+                }
+
+                return _sharedInstance;
+            }
+        }
+
+        /// <summary>
+        /// Creates a new ArrayPool instance of the given type.
+        /// </summary>
+        /// <param name="maxArrayLength">The maximum length of any Rent request.</param>
+        /// <param name-"numberOfArrays">The maximum number of arrays that will be rented.</param>
+        /// <returns>Returns a new ArrayPool<T> instance with the specified configuration options</return>
+        public static ArrayPool<T> Create(int maxArrayLength = DefaultArrayLength, int numberOfArrays = DefaultNumberOfArraysPerBucket)
+        {
+            return new DefaultArrayPool<T>(maxArrayLength, numberOfArrays);
+        }
+
+        /// <summary>
+        /// Gets the number of buffers that have been Rented and not yet Returned. This count includes
+        /// any buffers that have been allocated on-demand.
+        /// </summary>
+        public abstract int RentedBufferCount { get; }
+
+        /// <summary>
+        /// Gets the total number of allocated buffers available for Rent. This count includes any outstanding
+        /// buffers that have been allocated on-demand.
+        /// </summary>
+        public abstract int TotalBufferCount { get; }
+
+        /// <summary>
+        /// Retrieves a buffer from the pool that is at least the requested length. This buffer is loaned
+        /// to the caller and the caller must call Return when finished with the buffer.
+        /// </summary>
+        /// <param name="length">The number of elements in the desired buffer</param>
+        /// <returns>Returns an array of type T that is at least 'length' in size</returns>
+        public abstract T[] Rent(int length);
+
+        /// <summary>
+        /// Retrieves a buffer from the pool that is at least the requested new length. This buffer
+        /// contains all the data from the passed in buffer; the caller gives up ownership of the 
+        /// passed in buffer and must only use the new buffer reference.
```

Comment:
You need to add note that the data is copied from the old buffer to the new buffer.